tests/functional/conservation/conserv.py

```diff
@@ -0,0 +1,289 @@
+        return 0.5*np.sum((particles.v[:,0]**2)* particles.weights)
+
+
+
+def total_particles(parts, fun, lvlNbr=0, **kwargs):
+    """
+    given a particle hierarchy, calculate a total
+    quantity on a given level, quantity being estimated
+    by the callback "fun" (could be kinetic energy, momentum, et.)
+    """
+    for ilvl, lvl in parts.patch_levels.items():
+        if lvlNbr == ilvl:
+            tot = 0.
+            for ip, patch in enumerate(lvl.patches):
+                keys = list(patch.patch_datas.keys())
+                pdata = patch.patch_datas[keys[0]]
+                particles= pdata.dataset
+                per_patch = fun(particles, **kwargs)
+                tot += per_patch
+    return tot
+
+
+
+def total_kinetic(parts, lvlNbr=0, kind="iso"):
+    return total_particles(parts, kinetic_energy, lvlNbr, kind=kind)
+
+
+
+def mag_energy(B, lvlNbr=0):
+    """
+    return the total magnetic energy on a given level
+    """
+    for ilvl, lvl in B.patch_levels.items():
+        if lvlNbr == ilvl:
+            tot = 0.
+            for ip, patch in enumerate(lvl.patches):
+                pdata = patch.patch_datas["Bx"]
+
+                # Bx is primal, but By and Bz dual so
+                # we average By and Bz onto Bx to use them
+                # in the same formula
+                bx = patch.patch_datas["Bx"].dataset[5:-5]
+                bytmp = patch.patch_datas["By"].dataset[4:-4]
+                bztmp = patch.patch_datas["Bz"].dataset[4:-4]
+                by = 0.5*(bytmp[1:]+bytmp[:-1])
+                bz = 0.5*(bztmp[1:]+bztmp[:-1])
+
+                # sum 0.5B^2 * dx over all nodes
+                per_patch = np.sum((bx**2 + by**2 + bz**2)*0.5*pdata.layout.dl[0])
+                tot +=  per_patch
+        return tot
+
+
+def energies(path, kkind="iso"):
+    """
+    This loops over all times of a given run
+    and return the magnetic and kinetic energy
+    as a function of time.
+    """
+    r = Run(path)
+    times = get_times(r.path+"/EM_B.h5")
+    Bnrj = np.zeros_like(times)
+    K = np.zeros_like(times)
+    for it,t in enumerate(times):
+        B = r.GetB(t)
+        protons = r.GetParticles(t, "protons")
+        Bnrj[it] = mag_energy(B)
+        K[it] = total_kinetic(protons,kind=kkind)
+    return r, Bnrj, K, times
+
+
+
+def avg_interval(t1, t2, times):
+    return [np.argmin(np.abs(times-t)) for t in (t1,t2)]
+
+
+
+
+#############################################################
+#############################################################
+
+
+
+def main():
+
+    cases = [0.01,0.05,0.1,0.3,0.5,0.75,1,2]
+
+    dls = [0.2, 0.1]
+    nbrcells = [100,200]
+    nbrdts = [25000, 100000]
+
+    for vth in cases:
+        for dl, nbrcell, nbrdt in zip(dls, nbrcells, nbrdts):
+            uniform(vth, dl, nbrcell, nbrdt)
+            simulator = Simulator(gv.sim)
+            simulator.initialize()
+            simulator.run()
+            gv.sim = None
+
+
+
+    paths = glob("*vth*")
```

Comment:
What's happening here? 

Just
```
from glob import glob
glob("*vth*")
```

results in an empty list :\

Reply:
where? The runs diags are configured this way:
```
options": {"dir": "vth{}dx{}".format(vth,dl)
```
so that line should list all diag directories, no?

Reply:
ok it's directory listing with pattern matching, a bit different from what BASH considers "globbing" hence the confusion.

Reply:
maybe "wrong level of abstraction"

```
def directories_from_wildcard(pattern): 
    from glob import glob
    return glob(pattern)
```

or something perhaps